lib/astarte/client/housekeeping/realms.ex

```diff
@@ -0,0 +1,112 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2021 SECO Mind
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.Client.Housekeeping.Realms do
+  alias Astarte.Client.{APIError, Housekeeping}
+
+  def list(%Housekeeping{} = client) do
+    request_path = "realms"
+    tesla_client = client.http_client
+
+    with {:ok, %Tesla.Env{} = result} <- Tesla.get(tesla_client, request_path) do
+      if result.status == 200 do
+        {:ok, result.body}
+      else
+        {:error, %APIError{status: result.status, response: result.body}}
+      end
+    end
+  end
+
+  def create(%Housekeeping{} = client, realm_name, public_key_pem, opts)
+      when is_binary(realm_name) and is_binary(public_key_pem) and is_list(opts) do
+    request_path = "realms"
+    tesla_client = client.http_client
+
+    data = %{
+      realm_name: realm_name,
+      jwt_public_key_pem: public_key_pem
+    }
+
+    with {:ok, replication_data} <- fetch_replication(opts),
+         realm_data = Map.merge(data, replication_data),
+         {:ok, %Tesla.Env{} = result} <-
+           Tesla.post(tesla_client, request_path, %{data: realm_data}) do
+      if result.status == 201 do
+        :ok
+      else
+        {:error, %APIError{status: result.status, response: result.body}}
+      end
+    end
+  end
+
+  defp fetch_replication(opts) do
+    fetch_replication(opts, :datacenter_replication_factors)
+  end
+
+  defp fetch_replication(opts, :datacenter_replication_factors = key) do
+    case Keyword.fetch(opts, key) do
+      {:ok, datacenter_map} when is_map(datacenter_map) ->
+        %{
+          datacenter_replication_factors: datacenter_map,
+          replication_class: "NetworkTopologyStrategy"
+        }
+
+      :error ->
+        fetch_replication(opts, :replication_factor)
+    end
+  end
+
+  defp fetch_replication(opts, :replication_factor = key) do
+    case Keyword.fetch(opts, key) do
+      {:ok, replication_factor} ->
+        %{
+          replication_factor: replication_factor,
+          replication_class: "SimpleStrategy"
+        }
+
+      :error ->
+        {:error, :missing_replication_factor_and_datacenter_replication_factors}
+    end
+  end
```

Comment:
I'd flatten all this with a `cond` to make it more readable:
```elixir
defp fetch_replication(opts) do
  cond do
    Keyword.has_key?(opts, :datacenter_replication_factors) ->
      replication = %{
        datacenter_replication_factors: Keyword.fetch!(opts, :datacenter_replication_factors),
        replication_class: "NetworkTopologyStrategy"
      }
      {:ok, replication}

    Keyword.has_key?(opts, :replication_factor) ->
      ...

    true ->
      {:error, :missing_replication}
  end
end
```

Reply:
also, `:missing_replication` is enough (your error is way too long imho) and also if you return `{:error, reason}` you have to wrap a correct result in an `{:ok, result}` tuple (which you weren't doing)